rfcs/20190506-filesystem-plugin-modular-tensorflow.md

```diff
@@ -0,0 +1,829 @@
+
+```cpp
+class ModularRandomAccessFile : public RandomAccessFile {
+ public:
+  ModularRandomAccessFile(const string& filename,
+                          const TF_RandomAccessFile* file,
+                          const TF_RandomAccessFileOps* ops)
+      : filename_(filename), file_(file), ops_(ops) {}
+  virtual ~ModularRandomAccessFile() {
+    ops_->Cleanup(const_cast<TF_RandomAccessFile*>(file_));
+    delete file_;
+  }
+
+  Status Read(uint64 offset, size_t n, StringPiece* result, char* scratch) const;
+  Status Name(StringPiece* result) const;
+
+ private:
+  string filename_;
+  const TF_RandomAccessFile* file_;
+  const TF_RandomAccessFileOps* ops_;
+  TF_DISALLOW_COPY_AND_ASSIGN(ModularRandomAccessFile);
+};
+
+Status ModularRandomAccessFile::Read(uint64 offset, size_t n, StringPiece* result, char* scratch) const {
+  TF_Status* plugin_status = TF_NewStatus();
+  size_t read = ops_->Read(file_, offset, n, scratch, plugin_status);
+  *result = StringPiece(scratch, read);
+  Status status = StatusFromTF_Status(plugin_status);
+  TF_DeleteStatus(plugin_status);
+  return status;
+}
+
+Status ModularRandomAccessFile::Name(StringPiece* result) const {
+  *result = filename_;
+  return Status::OK();
+}
+```
+
+Similarly, we will have to implement a subclass of the `FileSystem` class,
+`ModularFileSystem`, in the same vein.
+
+Another change we have to do is to `FileSystemRegistry`: the `Register()` call
+after this proposal will use a `unique_ptr<FileSystem>` argument instead of the
+existing `Factory`-based implementation since the constructor of
+`ModularFileSystem` needs to take as argument the function tables for the
+operations supported by the filesystem.
+
+Because of this change, we also need a new subclass of `Env`, as each instance
+of `Env` has the `FileSystemRegistry` baked in at construction time. In fact, we
+can just change the existing class to have two lookup layers for the two
+registries.
+
+Furthermore, since most of the functionality calls `Env::Default()` to obtain
+the static instance of `Env` and then extract the filesystem from there, we can
+get rid of the level of indirection via `Env` and create a static instance of
+`ModularFileSystem`. This will be possible only after all filesystems use the
+modular approach, after the changes mentioned in the next section.
+
+Since the function tables in the filesystem interface mention this, we will have
+to move `TF_FileStatistics` from `tensorflow/c/env.h` to the same location as
+the filesystem plugin interface. Then, the outer level C API can just include
+this header and the structure will still be visible there.
+
+Once these changes are in place, the rest of the TensorFlow code can use the
+modular filesystem API transparently, with no change.
+
+### Optional additional changes to TensorFlow C/C++ core
+
+Something we might elect to do is to force all future filesystem implementations
+to use the plugin design suggested above. To do that, we will have to convert
```

Comment:
+1 let's definitely aim to have only one way of specifying filesystems in the future